File: DSA/Stack/Code/EvaluationOfInfixExpression.cs
// Evaluation of Infix Expression in C#

using System;
using System.Collections.Generic;

class EvaluationOfInfixExpression {
    static int Precedence(char c) {
        if (c == '+' || c == '-') return 1;
        if (c == '*' || c == '/') return 2;
        if (c == '^') return 3;
        return 0;
    }
    
    static int Evaluate(int a, int b, char op) {
        switch(op) {
            case '+': return a + b;
            case '-': return a - b;
            case '*': return a * b;
            case '/': return a / b;
        }
        return 0;
    }
    
    static int EvaluateInfixExpression(string infix) {
        Stack<int> nums = new Stack<int>();
        Stack<char> ops = new Stack<char>();
        
        foreach (char c in infix) {
            if (char.IsDigit(c)) {
                nums.Push(c - '0');
            } else if (c == '(') {
                ops.Push(c);
            } else if (c == ')') {
                while (ops.Count > 0 && ops.Peek() != '(') {
                    char op = ops.Pop();
                    int b = nums.Pop();
                    int a = nums.Pop();
                    nums.Push(Evaluate(a, b, op));
                }
                ops.Pop(); // Remove '('
            } else if (c == '+' || c == '-' || c == '*' || c == '/') {
                while (ops.Count > 0 && ops.Peek() != '(' && 
                       Precedence(ops.Peek()) >= Precedence(c)) {
                    char op = ops.Pop();
                    int b = nums.Pop();
                    int a = nums.Pop();
                    nums.Push(Evaluate(a, b, op));
                }
                ops.Push(c);
            }
        }
        
        while (ops.Count > 0) {
            char op = ops.Pop();
            int b = nums.Pop();
            int a = nums.Pop();
            nums.Push(Evaluate(a, b, op));
        }
        
        return nums.Pop();
    }
    
    static void Main() {
        string infix = "2+3*4-5";
        
        Console.WriteLine("=== Evaluation of Infix Expression ===\n");
        Console.WriteLine("Infix Expression: " + infix);
        Console.WriteLine("Result: " + EvaluateInfixExpression(infix) + "\n");
        
        Console.WriteLine("Operator Precedence: * / > + -");
        Console.WriteLine("Complexity: O(n)");
    }
}

File: DSA/Stack/Code/EvaluationOfInfixExpression.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX 100

// Evaluation of Infix Expression in C
typedef struct {
    int arr[MAX];
    int top;
} IntStack;

typedef struct {
    char arr[MAX];
    int top;
} CharStack;

IntStack* createIntStack() {
    IntStack* s = (IntStack*)malloc(sizeof(IntStack));
    s->top = -1;
    return s;
}

CharStack* createCharStack() {
    CharStack* s = (CharStack*)malloc(sizeof(CharStack));
    s->top = -1;
    return s;
}

void pushInt(IntStack* s, int val) {
    s->arr[++s->top] = val;
}

int popInt(IntStack* s) {
    return s->arr[s->top--];
}

void pushChar(CharStack* s, char c) {
    s->arr[++s->top] = c;
}

char popChar(CharStack* s) {
    return s->arr[s->top--];
}

char peekChar(CharStack* s) {
    return s->arr[s->top];
}

int isEmptyChar(CharStack* s) {
    return s->top == -1;
}

int precedence(char c) {
    if (c == '+' || c == '-') return 1;
    if (c == '*' || c == '/') return 2;
    if (c == '^') return 3;
    return 0;
}

int evaluate(int a, int b, char op) {
    switch(op) {
        case '+': return a + b;
        case '-': return a - b;
        case '*': return a * b;
        case '/': return a / b;
    }
    return 0;
}

int evaluateInfix(char* infix) {
    IntStack* nums = createIntStack();
    CharStack* ops = createCharStack();
    
    for (int i = 0; infix[i]; i++) {
        char c = infix[i];
        
        if (isdigit(c)) {
            pushInt(nums, c - '0');
        } else if (c == '(') {
            pushChar(ops, c);
        } else if (c == ')') {
            while (!isEmptyChar(ops) && peekChar(ops) != '(') {
                char op = popChar(ops);
                int b = popInt(nums);
                int a = popInt(nums);
                pushInt(nums, evaluate(a, b, op));
            }
            popChar(ops); // Remove '('
        } else if (c == '+' || c == '-' || c == '*' || c == '/') {
            while (!isEmptyChar(ops) && peekChar(ops) != '(' && 
                   precedence(peekChar(ops)) >= precedence(c)) {
                char op = popChar(ops);
                int b = popInt(nums);
                int a = popInt(nums);
                pushInt(nums, evaluate(a, b, op));
            }
            pushChar(ops, c);
        }
    }
    
    while (!isEmptyChar(ops)) {
        char op = popChar(ops);
        int b = popInt(nums);
        int a = popInt(nums);
        pushInt(nums, evaluate(a, b, op));
    }
    
    return popInt(nums);
}

int main() {
    char infix[MAX] = "2+3*4-5";
    
    printf("=== Evaluation of Infix Expression ===\n\n");
    printf("Infix Expression: %s\n", infix);
    printf("Result: %d\n\n", evaluateInfix(infix));
    
    printf("Operator Precedence: * / > + -\n");
    printf("Complexity: O(n)\n");
    
    return 0;
}
